OSDOCS-9480: NetObserv Custom Metrics (#74412)
openshift/openshift-docs: modules/network-observability-configuring-custom-metrics.adoc

// Module included in the following assemblies:
//

> // network_observability/metrics-alerts-dashboards.adoc

Author: @jotak originally, I had this procedure right after [Custom metrics ](https://github.com/openshift/openshift-docs/pull/74412/files#diff-0cb28d92ab17f03e9453aa0edd11dbd5d65d9017e0d4614c9e1d2a7ba42a3345)in the _Using metrics with dashboards and alerts_ section. I'm not 100% sold on it being in Configuring, but I tried moving it there and creating Additional resource links to see. It has a FlowCollector configuration piece, so in a way, it could indeed go in the Operator configuration. 

To give an argument for why we may consider moving it back to Using metrics with dashboards and alerts, many of the procedures in [Observing](https://docs.openshift.com/container-platform/4.12/observability/network_observability/observing-network-traffic.html) network traffic have FlowCollector configurations too, but we keep them in Observing because they tie in more with the web console views and they aren't really required configurations, similar to this one.

Reviewer: In fact these steps configuring FlowCollector are not necessary for using FlowMetrics, it's just that they enable new features that come with new fields in flow, and since the metrics API allows to gather data from any flow field, these new fields can be leveraged there. But users may also very well create custom metrics with any of the default/old flow fields

Author: Okay, so that gives me a good reason to keep the procedure in the Operator configuration then, since it isn't necessarily only for using FlowMetrics.

Author: On second thought, given that the FlowCollector steps aren't necesary any longer, I think I will move it back to the Metrics, alerts and dashboards page.

:_mod-docs-content-type: PROCEDURE
[id="network-observability-configuring-custom-metrics_{context}"]
= Configuring custom metrics
. In the web console, go to Operators → Installed Operators.
. Under the Provided APIs heading for the NetObserv Operator, select Flow Collector.
. Select cluster then select the YAML tab.
. Edit the `FlowCollector`  to add the following parameters for `spec.processor`:
  `processor.subnetLabels` field to create the FlowMetrics API.
+
[source,yaml]
—-
apiVersion: flows.netobserv.io/v1beta2
kind: FlowCollector
metadata:
  name: cluster
Spec:
# …
processor:
  addZone: true                <1>
  subnetLabels:
    openShiftAutoDetect: true  <2>
    customLabels:
    - cidrs: []
      name: ""
# …
—-
<1> Optional: Configure the ability to see zones in the metrics data. 

> <2> spec.processor.subnetLabels.openShiftAutoDetect` set to `true` to enable custom metrics. 

Reviewer: You can remove this part as we finally decided to enable this by default

Author: Removed

   direction: Ingress       <3>
   Labels: [SrcK8S_Name,SrcK8S_Namespace,SrcK8S_OwnerName,SrcK8S_OwnerType,DstAddr]       <4>
   Filters:                <5>
     - field: DstK8S_OwnerType
       matchType: Absence
     - field: SrcK8S_Type
       value: Pod
----
<1> How the metrics will appear, with the prefix `netobserv`. 
<2> Counter is useful for counting bytes or packets. Histogram is the other available type, which is useful for latencies. 
<3> Direction of traffic to capture. Ingress is an alternative option. If not specified, both ingress and egress are captured. 
<4> Labels define what the metrics look like and the relationship between the different entities. The labels also define the metrics cardinality. For example, `SrcK8S_Name` is a high cardinality metric.
<5> Filters allow further refinement of what you want to see. In this example, the filters ultimately refine to show external traffic, as the name indicates pods talking outside of the cluster. The `field: DstK8S_OwnerType` and `matchType: Absence` means if there is an absence of recognized IP address from the OwnerType, it is not a known ID, so there is a good chance that it is external. 
====
+

> .A-Z aware

Reviewer: Honestly, I'd prefer to avoid linking documentation and actual examples too closely, as this could become a maintenance burden. In fact if you look at [this pull request](https://github.com/netobserv/network-observability-operator/pull/609) that I have, it already makes the doc outdated 😬
Perhaps we can take just 2 examples, one showing a Counter metric and another a Histogram metric, as they are made slightly differently.

// Module included in the following assemblies:
//
// network_observability/metrics-alerts-dashboards.adoc

> :_mod-docs-content-type: PROCEDURE

Author: @jotak I think we may need to mention in steps somewhere here, if you haven't created a FlowMetric resource, you should create it. 
![image](https://github.com/openshift/openshift-docs/assets/77501157/f95958bd-c103-4bf1-9278-b19a046e7fdd)

Author: Are there any specific things to tweak in the form view for the FlowMetric API?

Author: Then you can use the YAML samples for the config of the `FlowMetric`?

Reviewer: exactly

[id="network-observability-configuring-custom-metrics_{context}"]
= Configuring custom metrics
. In the web console, navigate to the Import icon, +. 
. Here are a couple of examples of `FlowMetrics` configurations you can add:
+
.Generate a metric that tracks ingress bytes received from cluster external sources
[%collapsible]
====
[source,yaml]
----
apiVersion: flows.netobserv.io/v1alpha1
kind: FlowMetric
metadata:
  name: flowmetric-cluster-external-ingress-traffic
spec:
  metricName: cluster_external_ingress_bytes_total  <1>
  type: Counter                                     <2>
  valueField: Bytes
  direction: Ingress                                <3>
  labels: [DstK8S_HostName,DstK8S_Namespace,DstK8S_OwnerName,DstK8S_OwnerType] <4>
  filters:                                          <5>
  - field: SrcSubnetLabel
    matchType: Absence
----
<1> How the metrics appear, along with the prefix `netobserv`. 
<2> The `type` `Counter` is useful for counting bytes or packets. Histogram is the other available type, which is useful for latencies. 

> <3> Direction of traffic to capture. Egress is an alternative option. If not specified, both ingress and egress are captured. 

Reviewer: perhaps add something like:
"[...] are captured, which can lead to duplicated counts."

This is so users understand the reason for specifying a direction: it avoids duplication, since a single packet flowing from A to B will be reported once by A and once by B, leading in duplication. Specifying Ingress or Egress tells to report it only once. We could write a whole page explaining the subtelties of reporters & duplication, but trying here to tell just the essentials.. :-)

Author: Thanks joel. I agree its good to explain subtleties like this via examples in code callouts than in conceptual pieces.

> <3> Direction of traffic to capture. Ingress is an alternative option. If not specified, both ingress and egress are captured. 

Reviewer: Egress is the alternative, not Ingress.

```suggestion
<3> Direction of traffic to capture. Egress is an alternative option. If not specified, both ingress and egress are captured. 
```

Author: Addressed

> <4> Labels define what the metrics look like and the relationship between the different entities. The labels also define the metrics cardinality. For example, `SrcK8S_Name` is a high cardinality metric.

Reviewer: I think we need a warning or disclaimer here about high cardinality, telling that is can affect the memory usage of Prometheus.

On a related note, we have added this task https://issues.redhat.com/browse/NETOBSERV-1610 to add to the [flow format doc](https://docs.openshift.com/container-platform/4.15/observability/network_observability/json-flows-format-reference.html) whether each field is "suitable" for use in Prometheus (ie. is it high cardinality or not) => so perhaps would be good to link to the flow format reference here, to get more information about each field

> <5> Filters allow further refinement of what you want to see. In this example,  selecting only the cluster external traffic is done by matching only flows where `SrcSubnetLabel` is absent. This assumes the subnet labels feature is enabled (via `spec.processor.subnetLabels`) and configured to recognize IP ranges used in the cluster. In OpenShift, this is enabled and configured by default.

Reviewer: Removing mention "in OpenShift..." since it's openshift doc :-)

```suggestion
<5> Filters allow further refinement of what you want to see. In this example,  selecting only the cluster external traffic is done by matching only flows where `SrcSubnetLabel` is absent. This assumes the subnet labels feature is enabled (via `spec.processor.subnetLabels`) and configured, which is done by default.
```

  type: Histogram                 <2>
  valueField: TimeFlowRttNs
  direction: Ingress
  labels: [DstK8S_HostName,DstK8S_Namespace,DstK8S_OwnerName,DstK8S_OwnerType]
  filters:
  - field: SrcSubnetLabel
    matchType: Absence
  - field: TimeFlowRttNs
    matchType: Presence
  divider: "1000000000"      <3>
  buckets: [".001", ".005", ".01", ".02", ".03", ".04", ".05", ".075", ".1", ".25", "1"]  <4>
----
<1> The same namespace, `netobserv`, is used across all example components in this documentation. You can optionally use a different namespace for the different components. It is important to keep namespace names consistent across all components.
<2> The `type` here is `Histogram`, since it looks for a latency value (`TimeFlowRttNs`).
<3> Since the Rount-trip time (RTT) is provided as nanos in flows, use a divider of 1 billion to convert into seconds, which is standard in Prometheus guidelines.
<4> Custom buckets offer a decent precision on RTT, ranging approximately between 5ms and 250ms.
====

.Verification
. Once the pods refresh, Navigate to *Observe* -> *Metrics*.
. In the Expression field, you can type the metric name to see it displayed. 

> . You can also enter an expression, such as `topk(5, sum(rate(netobserv_pods_talking_outside_bytes_totall(ScK8S_Namespace="netobserv") [1m])) by

Reviewer 2: ```suggestion
. You can also enter an expression, such as `topk(5, sum(rate(netobserv_pods_talking_outside_bytes_total(ScK8S_Namespace="netobserv") [1m])) by
```

> . You can also enter an expression, such as `topk(5, sum(rate(netobserv_pods_talking_outside_bytes_totall(ScK8S_Namespace="netobserv") [1m])) by

Reviewer 2: we should probably stick to metric name that we demonstrate configuration for in above examples as opposed to `netobserv_pods_talking_outside_bytes_total`

Reviewer: yeah I guess that was the old name we used before, it should be changed to `netobserv_cluster_external_ingress_bytes_total`

The whole query (I checked it works) :

```
topk(5, sum(rate(netobserv_cluster_external_ingress_bytes_total{DstK8S_Namespace="my-namespace"}[2m])) by (DstK8S_HostName, DstK8S_OwnerName, DstK8S_OwnerType))
```

Author: Okay thanks. I changed the name in the expression, and I also relocated it to a `.Verification` step under the sample it refers to.

// Module included in the following assemblies:
//
// network_observability/metrics-alerts-dashboards.adoc

:_mod-docs-content-type: PROCEDURE
[id="network-observability-configuring-custom-metrics_{context}"]
= Configuring custom metrics using FlowMetrics API
. In the web console, navigate to *Operators* -> *Installed Operators*.

> . Under the *Provided APIs* heading for the *NetObserv Operator*, select *FlowMetrics*.

Reviewer 3: I am not sure how the UI looks like. You could call it something like `dropdown`, `section` or the UI element that you are pointing to instead of `heading`.
```suggestion
. In the *Provided APIs* heading for the *NetObserv Operator*, select *FlowMetrics*.
```

> . Use the *Project:*  dropdown list to select the project of the Network Observability Operator instance. 

Reviewer 3: ```suggestion
. In the *Project:*  dropdown list, select the project of the Network Observability Operator instance. 
```

. Click *Create FlowMetric*.

> . Here are a couple of examples of `FlowMetrics` configurations you can add:

Reviewer 3: This doesn't seem like a step of procedure. You may want to move it to a different section.

Author: Ok I can agree with this. I adjusted the step language to instead say: "Configure the `FlowMetrics` resource, such as the following examples" then there are a couple of examples you can use to add configurations.

> . Here are a couple of examples of `FlowMetrics` configurations you can add:

Reviewer 3: ```suggestion
.`FlowMetrics` configuration examples
```

+
.Generate a metric that tracks ingress bytes received from cluster external sources
[%collapsible]
====
[source,yaml]
----
apiVersion: flows.netobserv.io/v1alpha1
kind: FlowMetric
metadata:
  name: flowmetric-cluster-external-ingress-traffic
  namespace: netobserv                              <1>
spec:
  metricName: cluster_external_ingress_bytes_total  <2>
  type: Counter                                     <3>
  valueField: Bytes
  direction: Ingress                                <4>
  labels: [DstK8S_HostName,DstK8S_Namespace,DstK8S_OwnerName,DstK8S_OwnerType] <5>
  filters:                                          <6>
  - field: SrcSubnetLabel
    matchType: Absence
----

> <1> The same namespace, `netobserv`, is used across all example components in this documentation. You can optionally use a different namespace for the different components. It is important to keep namespace names consistent across all components.

Reviewer 3: I find the 2nd and 3rd sentences contradicting each other. Or may be this can be rephrased better. The first sentence can be skipped.

Reviewer 2: +1, we should simplify this.  thanks!

Author: Thanks, all :)

> <2> How the metrics appear, along with the prefix `netobserv`. 

Reviewer 3: The name of the metrics along with the prefix `netobserv`

Reviewer: I guess what we want to say here is that this is the name in Prometheus, and will be automatically prefixed with `netobserv_`.

Author: Ok I adjusted to say this: The name of the Prometheus metric, which in the web console appears with the prefix `netobserv-<metricName>`. Better, or not quite?

> <3> The `type` `Counter` is useful for counting bytes or packets. Histogram is the other available type, which is useful for latencies. 

Reviewer 3: You could begin with what is `type` followed by the types of `type`. If `Counter` is not a mandatory value here, use a placeholder instead

Author: GOod idea. I also think maybe I'll just leave the `Histogram` bits to the other example.

> <4> Direction of traffic to capture. Egress is an alternative option. If not specified, both ingress and egress are captured, which can lead to duplicated counts. 

Reviewer 3: You can use a placeholder instead.
```suggestion
<4> The direction of traffic to capture. The valid values are `Ingress` and `Egress`. If not specified, both ingress and egress are captured, which can lead to duplicated counts. 
```

Author: I think to simplify this, I'll skip the 2nd sentence of valid values, since Egress is used in the other example.

> <5> Labels define what the metrics look like and the relationship between the different entities. The labels also define the metrics cardinality. For example, `SrcK8S_Name` is a high cardinality metric.

Reviewer 3: May be rephrase to combine both into a single sentence. I do not have the technical context hence not sure if my suggestion would be accurate. Also, one question: Is cardinality a word that would need an explanation?

Reviewer: On cardinality, I can see that it's used in other parts of the documentation, without much explanation. [This page](https://docs.openshift.com/container-platform/4.15/support/troubleshooting/investigating-monitoring-issues.html#determining-why-prometheus-is-consuming-disk-space_investigating-monitoring-issues) mentions cardinality and shows a procedure for debugging high cardinality issues. I don't know, perhaps that would be good to point to that page for troubleshooting cardinality ?

Author: Okay I can link to this. I think thats a good idea, and good to point out that cardinality may not be widely discussed in the docs. Thanks both :)

> <6> Filters allow further refinement of what you want to see. In this example,  selecting only the cluster external traffic is done by matching only flows where `SrcSubnetLabel` is absent. This assumes the subnet labels feature is enabled (via `spec.processor.subnetLabels`), which is done by default.

Reviewer 3: `allow` is not allowed :D something like below

```suggestion
<6> Refines results based on the listed criteria. In this example,  selecting only the cluster external traffic is done by matching only flows where `SrcSubnetLabel` is absent. This assumes the subnet labels feature is enabled (via `spec.processor.subnetLabels`), which is done by default.
```

Author: Oh really, I don't find this in the IBMSG or the RHSSG? But I will take your rewrite anyways. I thought we could use allow, i.e. allowlist v. whitelist?

.Verification

> . Once the pods refresh, Navigate to *Observe* -> *Metrics*.

Reviewer 3: ```suggestion
. Once the pods refresh, navigate to *Observe* -> *Metrics*.
```

> . In the Expression field, you can type the metric name to see it displayed. 

Reviewer 3: something like below
```suggestion
. In the *Expression* field, type the metric name to view the corresponding result.
```

> . You can also enter an expression, such as `topk(5, sum(rate(netobserv_cluster_external_ingress_bytes_total{DstK8S_Namespace="my-namespace"}[2m])) by (DstK8S_HostName, DstK8S_OwnerName, DstK8S_OwnerType))`

Reviewer 3: This needn't be an independent step since it isn't an action. You can add this to the previous step.

====
+
.Show RTT latency for cluster external ingress traffic
[%collapsible]
====
[source,yaml]
----
apiVersion: flows.netobserv.io/v1alpha1
kind: FlowMetric
metadata:
  name: flowmetric-cluster-external-ingress-rtt
  namespace: netobserv    <1>
spec:
  metricName: cluster_external_ingress_rtt_seconds
  type: Histogram                 <2>
  valueField: TimeFlowRttNs
  direction: Ingress
  labels: [DstK8S_HostName,DstK8S_Namespace,DstK8S_OwnerName,DstK8S_OwnerType]
  filters:
  - field: SrcSubnetLabel
    matchType: Absence
  - field: TimeFlowRttNs
    matchType: Presence
  divider: "1000000000"      <3>
  buckets: [".001", ".005", ".01", ".02", ".03", ".04", ".05", ".075", ".1", ".25", "1"]  <4>
----

> <1> The same namespace, `netobserv`, is used across all example components in this documentation. You can optionally use a different namespace for the different components. It is important to keep namespace names consistent across all components.

Reviewer 3: Same comment as above

<2> The `type` here is `Histogram`, since it looks for a latency value (`TimeFlowRttNs`).

> <3> Since the Rount-trip time (RTT) is provided as nanos in flows, use a divider of 1 billion to convert into seconds, which is standard in Prometheus guidelines.

Reviewer 3: I am guessing it's `round`.
```suggestion
<3> Because the round-trip time (RTT) is provided as nanos in flows, specify a divider of 1 billion to convert into seconds, which is standard in Prometheus guidelines.
```

Author: Lol yes, the "t" in trip got ahead of me.

> <4> Custom buckets offer a decent precision on RTT, ranging approximately between 5ms and 250ms.

Reviewer 3: ```suggestion
<4> The custom buckets to specify precision on RTT. The values range between 5ms and 250ms.
```

Reviewer: The values don't range only between 5ms and 250ms, but the precision is better within this range

Author: OK. I updated to the following:
The custom buckets specify precision on RTT, with optimal precision ranging between 5ms and 250ms.

.Verification

> . Once the pods refresh, Navigate to *Observe* -> *Metrics*.

Reviewer 3: ```suggestion
. Once the pods refresh, navigate to *Observe* -> *Metrics*.
```

> . In the Expression field, you can type the metric name to see it displayed. 

Reviewer 3: something like below.
```suggestion
. In the *Expression* field, type the metric name to view the corresponding result.
```
